ic-http-mock/src/mock.rs

```diff
@@ -0,0 +1,134 @@
+pub fn mock(request: &CanisterHttpRequestArgument, response: &HttpResponse) {
+    mock_with_delay(request, response, Duration::from_secs(0));
+}
+
+/// Adds a mock for a given HTTP request and response. The mock will be returned by
+/// subsequent calls to `http_request` that match the same request. The response will be
+/// returned after a delay specified by the `delay` argument.
+pub fn mock_with_delay(
+    request: &CanisterHttpRequestArgument,
+    response: &HttpResponse,
+    delay: Duration,
+) {
+    insert(Mock {
+        request: request.clone(),
+        response: response.clone(),
+        delay,
+        times_called: 0,
+    });
+}
+
+/// Handles incoming HTTP requests by retrieving a mock response based
+/// on the request, possibly delaying the response, transforming the response if necessary,
+/// and returning it. If there is no mock found, it returns an error.
+pub async fn http_request(
+    request: &CanisterHttpRequestArgument,
+) -> Result<(HttpResponse,), (RejectionCode, String)> {
+    let mut mock =
+        get(request).ok_or((RejectionCode::CanisterReject, "No mock found".to_string()))?;
+    mock.times_called += 1;
+    insert(mock.clone());
+
+    // Delay the response if necessary.
+    if mock.delay > Duration::from_secs(0) {
+        // TODO: Use a non-blocking sleep when the CDK supports it.
+        #[cfg(not(target_arch = "wasm32"))]
+        tokio::time::sleep(mock.delay).await;
+    }
+
+    // Check if the response body exceeds the maximum allowed size.
+    if let Some(max_response_bytes) = mock.request.max_response_bytes {
+        if mock.response.body.len() as u64 > max_response_bytes {
+            return Err((
+                RejectionCode::SysFatal,
+                format!(
+                    "Value of 'Content-length' header exceeds http body size limit, {} > {}.",
+                    mock.response.body.len(),
+                    max_response_bytes
+                ),
+            ));
+        }
+    }
+
+    // Apply the transform function if one is specified.
+    let transform = match request.transform {
+        Some(ref t) => crate::transform::registry_get(t.function.0.method.clone()),
+        None => None,
+    };
+    let transformed_response = match transform {
+        Some(function) => function(TransformArgs {
+            response: mock.response.clone(),
+            context: vec![],
+        }),
+        None => mock.response.clone(),
+    };
+
+    Ok((transformed_response,))
+}
+
+/// Returns the number of times the given request has been called.
+/// Returns 0 if no mock has been found for the request.
+pub fn times_called(request: &CanisterHttpRequestArgument) -> u64 {
+    get(request).map(|mock| mock.times_called).unwrap_or(0)
+}
+
+/// Create a hash from a CanisterHttpRequestArgument, which includes its URL,
+/// method, headers, body, and optionally, its transform function name.
+fn hash(request: &CanisterHttpRequestArgument) -> String {
```

Comment:
This exists because `Hash` isn't derived for `CanisterHttpRequestArgument`, right?

Reply:
Yes, that's exactly right.